### PR TITLE
gnrc_netif: only use prefix matching as tie-breaker in source selection

### PR DESCRIPTION
**Pull request by miri64**
### Contribution description
When source address selection is done, both [RFC](https://tools.ietf.org/html/rfc6724#section-5) and comments in the code state, that a longest prefix match should *only* be used as a tie-breaker between more than one viable candidate. If there is only one address, there is

a) no need for a tie-breaker
b) in the case of either the destination address or the single remaining address being ULAs ([which are considered to be of global scope][RFC4193]) possibly not matching, as `fd00::/7` and e.g. `2001::/8` do not have a common prefix.

(b) in fact causes the match function to return -1, causing the source address selection to return -1, causing the outer function to return the first address it found (which most often is the link-local address), causing e.g. a ping to an ULA to fail, even is there is a global address.

I also piggy-backed a reduction of code duplication (so I can set `idx` just at one point) and added some doc on candidate creation, because some things were unclear to me when I tried to wrap my head around the problem.

[RFC4193]: https://tools.ietf.org/html/rfc4193

### Testing procedure
- Pinging the link-local address between two native instances should still select the link-local address of the sender (I tested with `examples/gnrc_networking` on `native` and sniffed `tapbr0` while pinging):
  ```
  ifconfig 6 add 2001:db8::1
  ping6 fe80::d826:95ff:fe26:f1d5
  ```
- When merging #12404, pinging `fd00:dead:beef::1` from a `examples/gnrc_networking` node via an `examples/gnrc_border_router` node should still work (after the global addresses are set up on all nodes).
- `make -C tests/gnrc_netif flash test` should still work for a board of your choice and should not work if you revert the fix to the prefix matching:
  - `git revert HEAD~2; git revert HEAD~4; make -C tests/gnrc_netif`

### Issues/PRs references
~~Fixes an issue in source address selection with ULAs becoming obvious in https://github.com/RIOT-OS/RIOT/pull/12404. Required to make https://github.com/RIOT-OS/RIOT/pull/12404 mergable.~~ (#12404 now is included here)

Requires #12425 for address resolution to still work fully (and `tests/gnrc_ipv6_nib` to pass)

### CODE REVIEW

**Comment by miri64** [comment]
After thinking about this some more, I think using `_match_to_idx()` in this function is completely the wrong approach, as if we have e.g. two global addresses starting with `0000::/1` assigned to the interface that win the source address selection so far and the destination address starts with `1000::/1` it will return `-1` as none of the addresses selected matches the destination address by more than 1 bit, causing the source address selection to pick `first_candidate` (which is typically is the auto-generated link-local address). Rather, we should compare the best matches of the remaining addresses with the destination, as rule 8 states, and pick the best (or either one of them if they match equally best, even if it is 0).

**Comment by miri64** [comment]
Done.

File: sys/net/gnrc/netif/gnrc_netif.c [hunk 1]
```diff
@@ -866,10 +861,7 @@ static int _match_to_idx(const gnrc_netif_t *netif,
     for (int i = 0; i < GNRC_NETIF_IPV6_ADDRS_NUMOF; i++) {
         unsigned match;
 
-        if ((netif->ipv6.addrs_flags[i] == 0) ||
-            ((filter != NULL) && _addr_anycast(netif, i)) ||
```
Comment (miri64) [inline sys/net/gnrc/netif/gnrc_netif.c]: See [RFC 6724, page 30](https://tools.ietf.org/html/rfc6724#page-30), point 3: This can be removed without alternative.

**Comment by miri64** [comment]
I added a test for the corner case, that popped up due to the fix in #12404.

**Comment by miri64** [comment]
Squashed, rebased, added the commit from #12404 and added some more test cases.

**Comment by miri64** [comment]
@fjmolinas @kb2ma can one of you maybe give this a look?

**Comment by miri64** [comment]
Pushed a bit too early and added some fixes to that function now.

**Comment by miri64** [comment]
Sorry, yet another issue... I'll just say when I'm done... :(

**Comment by kb2ma** [comment]
No worries, @miri64. I need a little time to understand the issues anyway.

**Comment by miri64** [comment]
Section 2.2 in RFC 6724 actually makes one of the test cases invalid, so I removed it. The tests are now passing again. Sorry for these last minute changes.

**Comment by miri64** [comment]
@kb2ma do you mind if I squash? I somehow messed up my fixup commits, I just noticed...

**Comment by kb2ma** [comment]
Go right ahead. I would likely skip over the intermediate results anyway.

**Comment by miri64** [comment]
Squashed

> Go right ahead. I would likely skip over the intermediate results anyway.

At least for the review I think it helps if you go through it commit-wise (not all tests work for all commits though), as I tried to make my thinking process clearer with the history. Additionally, the PR contains some clean-up of now unnecessary code, that resulted from fixing the bugs and adapting it more in line with RFC 6724.

**Comment by miri64** [comment]
(I'm now done, btw)

**Comment by kb2ma** [comment]
Sounds good. I've got a basic handle on the issues. Is the testing procedure in the PR description still valid?

**Comment by miri64** [comment]
Yepp

**Comment by miri64** [comment]
Except for the merging of #12404 part in (2), as that one is part of this PR now ;-).

**Comment by miri64** [comment]
Ah, it's `tests/gnrc_ipv6_nib` that fails... Will see what's going on.

**Comment by miri64** [comment]
The last commit should fix that. @kb2ma the reverts in the testing description moved by one commit backwards because of that.

**Comment by miri64** [comment]
>     * Is the gnrc_networking board a physical board?

Yes.
 
>     * Is it communicating wirelessly with the gnrc_border_router?

Yes.

>     * Are you using the 2001:db8 prefix for all boards?

The border router example should configure it for all available boards (so `gnrc_networking`) as soon as it comes up (might take a while to reach the `gnrc_networking` board, but resetting the latter speads things up)

>     * How is the host for the border router (I assume a linux workstation) configured?

As the `gnrc_border_router` example sets things up, once you hit `term`, so a `tap0` interface to which `2001:db8::/64` is routed and an ULA configured for the loopback interface.

> My experienced with global IPv6 addresses is limited. I'm used to using ULAs.

The examples should bootstrap everything for you.

**Comment by miri64** [comment]
(I used two samr21-xpro btw)

**Comment by miri64** [comment]
I post-poned this to the next release for 2 reasons:

- I don't like that I have to unlock the NIB mid-transaction now
- While this is a bug, it only occurs in certain corner cases

We should just note it down as a known issue for this release and I try to figure out how to do this without unlocking the NIB probing.

**Comment by kb2ma** [comment]
Sounds good to me.

**Comment by kb2ma** [comment]
Moved the network setup discussion to #12422.

**Comment by miri64** [comment]
Rebased to current master, no longer waiting for other PR.

File: sys/net/gnrc/netif/gnrc_netif.c [hunk 2]
```diff
@@ -955,8 +938,19 @@ static int _create_candidate_set(const gnrc_netif_t *netif,
...
         if ((netif->ipv6.addrs_flags[i] == 0) ||
+            /* https://tools.ietf.org/html/rfc4862#section-2:
+             *  A tentative address is not considered assigned to an interface
+             *  in the usual sense.  An interface discards received packets
+             *  addressed to a tentative address, but accepts Neighbor Discovery
+             *  packets related to Duplicate Address Detection for the tentative
+             *  address.
+             *  (so don't consider tentative addresses for source address
+             *  selection) */
             gnrc_netif_ipv6_addr_dad_trans(netif, i)) {
```
Comment (cgundogan) [inline sys/net/gnrc/netif/gnrc_netif.c]: I know this line is not part of this PR, but could you clarify why addresses are not considered that have `gnrc_netif_ipv6_addr_dad_trans() != 0`? According to the doc of this function, the return value is:
```
the number of duplicate address detection transmissions already performed
```
How can we deduce that the address is tentative from that result?

Comment (miri64) [reply sys/net/gnrc/netif/gnrc_netif.c]: The number of DAD transmissions is encoded in the tentative state in our DAD implementation (that's why you see `TNT[1..3]` with the `ifconfig` command, not just `TNT`). So a value of 0 for already performed DAD transmissions means the address is *not* tentative. As the NS for DAD is sent as soon as the address is added, this is also true if you don't have the knowledge of such GNRC internals ;)

Comment (miri64) [reply sys/net/gnrc/netif/gnrc_netif.c]: See also doc for tentative state:

https://github.com/RIOT-OS/RIOT/blob/55adbee48808bda975dd5ee420c605bbddf4c94c/sys/include/net/gnrc/netif/ipv6.h#L48-L55

**Comment by fjmolinas** [comment]
> `git revert HEAD~2; git revert HEAD~4; make -C tests/gnrc_netif` 

@miri64 just to be sure after new commits have been added, can you point out the hash of the two commits to revert?

**Comment by miri64** [comment]
> > `git revert HEAD~2; git revert HEAD~4; make -C tests/gnrc_netif`
> 
> @miri64 just to be sure after new commits have been added, can you point out the hash of the two commits to revert?

Sorry only noticed the activity in this PR now. I'll have a look at your review and then update :-).

File: sys/net/gnrc/netif/gnrc_netif.c [hunk 3]
```diff
@@ -1077,18 +1099,29 @@ static ipv6_addr_t *_src_addr_selection(gnrc_netif_t *netif,
...
+            /* if match == 0 for all case, it takes above selected idx */
+            if (match > best_match) {
+                idx = i;
+                best_match = match;
+            }
         }
     }
-    /* otherwise apply rule 8: Use longest matching prefix. */
-    int idx = _match_to_idx(netif, dst, candidate_set);
     return (idx < 0) ? NULL : &netif->ipv6.addrs[idx];
```
[the remark this answers is not in the document]
Comment (miri64) [reply sys/net/gnrc/netif/gnrc_netif.c]: 1. your suggestion dangerous to do this without checking `idx` first. Should I put the check around the debug as well?
```suggestion
    if (idx < 0) {
        DEBUG("No winner found\n);
        return NULL;
    }
    else {
         DEBUG("Winner is: %s\n",
             ipv6_addr_to_str(addr_str, &netif->ipv6.addrs[idx], sizeof(addr_str)));
         return &netif->ipv6.addrs[idx];
```
2. there is already an output of the source at least here https://github.com/RIOT-OS/RIOT/blob/d7c4eec91566c3bbcb74ae29a02c9bc5ab609181/sys/net/gnrc/network_layer/ipv6/gnrc_ipv6.c#L369-L377 (this is insufficient however of course if for other calls of this function or if you just want to debug this function isolated)...

How shall we proceed?

File: sys/net/gnrc/netif/gnrc_netif.c [hunk 4]
```diff
@@ -1077,18 +1099,29 @@ static ipv6_addr_t *_src_addr_selection(gnrc_netif_t *netif,
...
+    uint8_t best_match = 0;
     /* collect candidates with maximum points */
     for (int i = 0; i < GNRC_NETIF_IPV6_ADDRS_NUMOF; i++) {
         if (winner_set[i] == max_pts) {
-            bf_set(candidate_set, i);
+            const ipv6_addr_t *addr = &netif->ipv6.addrs[i];
+            unsigned match = ipv6_addr_match_prefix(addr, dst);
+
+            match = _cap_match(netif, addr, match);
+            /* if match == 0 for all case, it takes above selected idx */
```
[the remark this answers is not in the document]
Comment (miri64) [reply sys/net/gnrc/netif/gnrc_netif.c]: Yepp. But apart from the corner-case you face with CoAP sockets, this is not wrong, as both are valid source addresses.

**Comment by miri64** [comment]
> > `git revert HEAD~2; git revert HEAD~4; make -C tests/gnrc_netif`
> 
> @miri64 just to be sure after new commits have been added, can you point out the hash of the two commits to revert?

Currently `HEAD~2` would be fb1754e69016f2e6f85fbf6a41f5b357b9e555eb and `HEAD~4` d8f6d8270d4a726a55b61556adeeb8356f6f08e2.

File: sys/net/gnrc/netif/gnrc_netif.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (fjmolinas) [reply sys/net/gnrc/netif/gnrc_netif.c]: I would rather go with solution 1, that way I can only enable `DEBUG` in one file if that is Ok with you.

File: sys/net/gnrc/netif/gnrc_netif.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (fjmolinas) [reply sys/net/gnrc/netif/gnrc_netif.c]: The issue was that before it was selecting different addresses sometimes. As long as it is deterministic I can deal with it :)

**Comment by fjmolinas** [comment]
> not work if you revert the fix to the prefix matching:
> - git revert HEAD~2; git revert HEAD~4; make -C tests/gnrc_netif

completed test procedure...

```
flags: 0x0
src_l2addr: 3E:E6:B5:22:FD:0B
dst_l2addr: 3E:E6:B5:22:FD:0A
~~ PKT    -  2 snips, total size:  61 byte
Timeout in expect script at "child.expect(r"OK \(\d+ tests\)")" (tests/gnrc_netif/tests/01-run.py:16)
  File "/home/francisco/workspace/RIOT/dist/pythonlibs/testrunner/__init__.py", line 29, in run
    testfunc(child)
  File "/home/francisco/workspace/RIOT/tests/gnrc_netif/tests/01-run.py", line 16, in testfunc
    child.expect(r"OK \(\d+ tests\)")
  File "/home/francisco/.local/lib/python3.6/site-packages/pexpect/spawnbase.py", line 341, in expect
    timeout, searchwindowsize, async_)
  File "/home/francisco/.local/lib/python3.6/site-packages/pexpect/spawnbase.py", line 369, in expect_list
    return exp.expect_loop(timeout)
  File "/home/francisco/.local/lib/python3.6/site-packages/pexpect/expect.py", line 119, in expect_loop
    return self.timeout(e)
  File "/home/francisco/.local/lib/python3.6/site-packages/pexpect/expect.py", line 82, in timeout
    raise TIMEOUT(msg)

/home/francisco/workspace/RIOT/tests/gnrc_netif/../../Makefile.include:710: recipe for target 'test' failed
```

**Comment by miri64** [comment]
Note, because I was confused for a moment: The test is supposed to fail...

**Comment by fjmolinas** [comment]
@miri64 please squash!

**Comment by fjmolinas** [comment]
> @miri64 please squash!

Lets trigger tests in murdock as well!

**Comment by miri64** [comment]
Squashed.

**Comment by miri64** [comment]
Removed the empty commit that happened somehow due to squashing the revert commit on the original.

**Comment by miri64** [comment]
Thanks for the review!

**Comment by fjmolinas** [comment]
> Thanks for the review!

Thanks for the fix!
